spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -193,47 +193,438 @@ For example, calling an `@Outgoing` annotated method from user code will not pos
+| This method is called for every incoming message. Never concurrently, so the previous call must have returned before calling it with the next incoming _payload_.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+CompletionStage<Message<O>> method(Message<I> msg)
+----
+| Produces a `Message` for each incoming `Message`. This method returns a `CompletionStage` that can redeem the `Message` instance asynchronously. The returned `CompletionStage` must not be completed with `null`.
+| This method is called for every incoming message. Never concurrently. The reactive messaging implementation must wait until the completion of the previously returned `CompletionStage` before calling the method again with the next `Message`.
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+CompletionStage<O> method(I payload)
+----
+| Produces a _payload_ for each incoming _payload_. This method returns a `CompletionStage` that can redeem the _payload_ instance asynchronously. The returned `CompletionStage` must not be completed with `null`.
+| This method is called for every incoming _payload_. Never concurrently. The reactive messaging implementation must wait until the completion of the previously returned `CompletionStage` before calling the method again with the next _payload_.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Publisher<Message<O>> method(Publisher<Message<I>> pub)
+----
+| Transforms the incoming `Publisher` of `Message`. This method is used to manipulate streams and apply stream transformations.
+| This method is called once, at _assembly_ time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+PublisherBuilder<Message<O>> method(PublisherBuilder<Message<I>> pub)
+----
+| Transforms the incoming `PublisherBuilder` of `Message`. This method is used to manipulate streams and apply stream transformations.
+| This method is called once, at _assembly_ time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Publisher<O> method(Publisher<I> pub)
+----
+| Transforms the incoming `Publisher` of _payload_. This method is used to manipulate streams and apply stream transformations.
+| This method is called once, at _assembly_ time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+PublisherBuilder<O> method(PublisherBuilder<I> pub)
+----
+| Transforms the incoming `PublisherBuilder` of _payload_. This method is used to manipulate streams and apply stream transformations.
+| This method is called once, at _assembly_ time.
+
+|===
+
+===== Examples of simple method streams
 
 The simplest shape that an application may use is a simple method. This is a method that accepts an incoming message, and possibly publishes an outgoing message:
 
 [source, java]
 ----
-@Incoming
-@Outgoing
+@Incoming("in")
+@Outgoing("out")
 public OutgoingMessage processMessage(IncomingMessage msg) {
   return convert(msg);
 }
 ----
 
```

Comment:
Are the type names used IncomingMessage and OutgoingMessage slightly confusing when what is
actually being passed in is a _payload_? Suggest IncomingType and OutgoingType ?

Reply:
Or do we mean IncomingMessage implements Message<Type>?

Reply:
It meant that `IncomingMessage` was implementing `Message`, but I've rewritten this section.